python/tskit/ms.py

```diff
@@ -0,0 +1,114 @@
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Write the genotypes from tree sequences to ms-style output
+"""
+
+
+class msWriter:
+    """
+    Writes a ms representation of the genotypes in tree sequence to a
+    file-like object.
+    """
+
+    def __init__(
+        self,
+        tree_sequence,
+        print_trees=False,
+        precision=4,
+        mutation_rate=0,
+        num_replicates=1,
+        write_header=True,
+    ):
+        self._tree_sequence = tree_sequence
+        self._num_replicates = num_replicates
+        self._mutation_rate = mutation_rate
+        self._precision = precision
+        self._print_trees = print_trees
+        self._write_header = write_header
+
+    def print_trees(self, output):
+        """
+        Print out the trees in ms-format from the specified tree sequence.
+        When 'invisible' recombinations occur ms prints out copies of the
+        same tree. Therefore, we must keep track of all breakpoints from the
+        simulation and write out a tree for each one.
+        """
+        tree = next(self._tree_sequence.trees())
+        newick = tree.newick(precision=self._precision)
+        print(f"[{tree.span}]", newick, file=output)
+
+    def __write_header(self, output):
+        print(
+            "ms {} {}".format(
+                self._tree_sequence.sample_size, max(self._num_replicates, 1)
+            ),
+            file=output,
+        )
+        print("{}".format(999), file=output)
+
+    def write(self, output):
+
+        if self._write_header is True:
+            self.__write_header(output)
+
+        tree_sequence = self._tree_sequence
+        print(file=output)
+        print("//", file=output)
+        if self._print_trees is True:
+            self.print_trees(output)
+        if self._mutation_rate > 0:
+            s = tree_sequence.get_num_sites()
+            print("segsites:", s, file=output)
+            if s != 0:
+                print("positions: ", end="", file=output)
+                positions = [
+                    variant.position / (tree_sequence.sequence_length)
+                    for variant in tree_sequence.variants()
+                ]
+                for position in positions:
+                    print(
+                        "{0:.{1}f}".format(position, self._precision),
```

Comment:
You can use an f string here.

Reply:
I've replaced the printing with f strings.